### PR TITLE
feat: announcer adds announcerToTrainer

### PR DESCRIPTION
**Pull request by MinH-09**
Trainer in scheduler is responsible for transferring data to trainer for training model
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

File: scheduler/trainer/trainer_test.go [hunk 1]
```diff
@@ -0,0 +1,453 @@
...
+func Test_New(t *testing.T) {
+	tests := []struct {
+		name   string
+		mock   func(m *trainerclientmocks.MockV1MockRecorder, stream trainerv1.Trainer_TrainClient)
+		expect func(t *testing.T, tr Trainer, err error)
+	}{
+		{
+			name: "get trainerStream fail",
+			mock: func(m *trainerclientmocks.MockV1MockRecorder, stream trainerv1.Trainer_TrainClient) {
+				gomock.InOrder(
```
Comment (fcgxz2003) [inline scheduler/trainer/trainer_test.go:253]: Only one mock that does not require gomock.Inorder.

File: scheduler/trainer/trainer.go [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+
+// sendDataToTrainer send dataset to trainer for training model.
+func (t *trainer) sendDataToTrainer() error {
+	scheduler, err := t.dynconfig.GetScheduler()
+	if err != nil {
+		return err
+	}
+
+	var b bytes.Buffer
+	download, err := t.storage.ListDownload()
```
Comment (fcgxz2003) [inline scheduler/trainer/trainer.go:102]: downloads

File: scheduler/trainer/trainer.go [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+				TrainMlpRequest: &trainerv1.TrainMLPRequest{
+					Dataset: b.Bytes(),
+				},
+			},
+		}); err != nil {
+			return err
+		}
+	}
+
+	networkTopology, err := t.storage.ListNetworkTopology()
```
Comment (fcgxz2003) [inline scheduler/trainer/trainer.go:125]: networkTopologies

File: scheduler/trainer/trainer.go [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+	if err != nil {
+		return err
+	}
+
+	var b bytes.Buffer
+	download, err := t.storage.ListDownload()
+	if err != nil {
+		return err
+	}
+	for _, data := range download {
```
Comment (fcgxz2003) [inline scheduler/trainer/trainer.go:106]: for _, download := range downloads {

File: scheduler/trainer/trainer.go [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+		}); err != nil {
+			return err
+		}
+	}
+
+	networkTopology, err := t.storage.ListNetworkTopology()
+	if err != nil {
+		return err
+	}
+	for _, data := range networkTopology {
```
Comment (fcgxz2003) [inline scheduler/trainer/trainer.go:129]: for _, networkTopology := range networkTopologies {

File: scheduler/trainer/trainer_test.go [hunk 6]
```diff
@@ -0,0 +1,453 @@
...
+				m.Train(gomock.Any()).Return(stream, nil)
+				md.GetScheduler().Return(&managerv2.Scheduler{
+					Hostname: "foo",
+					Ip:       "127.0.0.1",
+					SchedulerCluster: &managerv2.SchedulerCluster{
+						Id: 1,
+					},
+				}, nil)
+				ms.ListDownload().Return([]storage.Download{mockDownload}, nil)
+				mts.Send(gomock.Any()).DoAndReturn(
```
Comment (fcgxz2003) [inline scheduler/trainer/trainer_test.go:420]: delete

File: scheduler/trainer/trainer_test.go [hunk 7]
```diff
@@ -0,0 +1,453 @@
...
+					func(t *trainerv1.TrainRequest) error {
+						return nil
+					})
+				ms.ListNetworkTopology().Return([]storage.NetworkTopology{mockNetworkTopology}, nil)
+				mts.Send(gomock.Any()).DoAndReturn(
```
Comment (fcgxz2003) [inline scheduler/trainer/trainer_test.go:425]: .AnyTimes()

File: scheduler/trainer/trainer_test.go [hunk 8]
```diff
@@ -0,0 +1,453 @@
...
+			except: func(t *testing.T, tr Trainer, err error) {
+				assert := assert.New(t)
+				assert.EqualError(err, "foo")
+			},
+		},
+		{
+			name: "list download fail",
+			mock: func(stream trainerv1.Trainer_TrainClient, m *trainerclientmocks.MockV1MockRecorder, md *configmocks.MockDynconfigInterfaceMockRecorder, ms *storagemocks.MockStorageMockRecorder, mts *trainerv1mocks.MockTrainer_TrainClientMockRecorder) {
+				m.Train(gomock.Any()).Return(stream, nil)
+				md.GetScheduler().Return(nil, nil)
```
Comment (fcgxz2003) [inline scheduler/trainer/trainer_test.go:312]: md.GetScheduler().Return(&managerv2.Scheduler{
					Hostname: "foo",
					Ip:       "127.0.0.1",
					SchedulerCluster: &managerv2.SchedulerCluster{
						Id: 1,
					},
				}, nil)

File: scheduler/announcer/announcer.go [hunk 9]
```diff
@@ -78,6 +104,27 @@ func (a *announcer) Serve() error {
 		return err
 	}
 
+	if a.config.Trainer.Enable {
```
Comment (gaius-qi) [inline scheduler/announcer/announcer.go:107]: Remove.

**Review by gaius-qi — CHANGES_REQUESTED** [review]

File: scheduler/announcer/announcer.go [hunk 10]
```diff
@@ -99,3 +134,91 @@ func (a *announcer) announceToManager() error {
...
+	if err != nil {
+		return err
+	}
+	defer downloadReadCloser.Close()
+	for {
+		n, err := downloadReadCloser.Read(buffer)
+		if err != nil && err != io.EOF {
+			return err
+		}
+		if err == io.EOF || n == 0 {
```
Comment (fcgxz2003) [inline scheduler/announcer/announcer.go:176]: remove n==0

File: scheduler/announcer/announcer.go [hunk 11]
```diff
@@ -99,3 +134,91 @@ func (a *announcer) announceToManager() error {
...
+	if err != nil {
+		return err
+	}
+	defer networkTopologyReadCloser.Close()
+	for {
+		n, err := networkTopologyReadCloser.Read(buffer)
+		if err != nil && err != io.EOF {
+			return err
+		}
+		if err == io.EOF || n == 0 {
```
Comment (fcgxz2003) [inline scheduler/announcer/announcer.go:203]: remove n==0

File: scheduler/announcer/announcer.go [hunk 12]
```diff
@@ -99,3 +129,91 @@ func (a *announcer) announceToManager() error {
...
+	}
+	defer networkTopologyReadCloser.Close()
+	for {
+		n, err := networkTopologyReadCloser.Read(buffer)
+		if err != nil && err != io.EOF {
+			return err
+		}
+		if n == 0 {
+			break
+		}
```
[the remark this answers is not in the document]
Comment (fcgxz2003) [reply scheduler/announcer/announcer.go:205]: `if err == io.EOF {
	break
}
if err != nil {
	return err
}`

File: scheduler/announcer/announcer.go [hunk 13]
```diff
@@ -99,3 +129,91 @@ func (a *announcer) announceToManager() error {
...
+	}
+	defer downloadReadCloser.Close()
+	for {
+		n, err := downloadReadCloser.Read(buffer)
+		if err != nil && err != io.EOF {
+			return err
+		}
+		if n == 0 {
+			break
+		}
```
[the remark this answers is not in the document]
Comment (fcgxz2003) [reply scheduler/announcer/announcer.go:178]: `if err == io.EOF {
	break
}
if err != nil {
	return err
}`

File: scheduler/scheduler.go [hunk 14]
```diff
@@ -78,6 +79,9 @@ type Server struct {
 	// Security client.
 	securityClient securityclient.V1
 
+	// trainer client.
```
Comment (zhoulizhen) [inline scheduler/scheduler.go:82]: // Trainer client.

**Review by fcgxz2003 — COMMENTED** [review]
LGTM

**Review by MIchaelFU0403 — COMMENTED** [review]
LGTM

File: scheduler/announcer/announcer.go [hunk 15]
```diff
@@ -99,3 +134,96 @@ func (a *announcer) announceToManager() error {
 
 	return nil
 }
+
+// announceToTrainer announces regularly dataset to trainer for training model.
+func (a *announcer) announceToTrainer() {
```
Comment (gaius-qi) [inline scheduler/announcer/announcer.go:139]: https://github.com/dragonflyoss/Dragonfly2/blob/main/client/daemon/announcer/announcer.go

**Review by gaius-qi — CHANGES_REQUESTED** [review]

**Review by fcgxz2003 — COMMENTED** [review]
LGTM

File: scheduler/announcer/announcer.go [hunk 16]
```diff
@@ -73,6 +105,11 @@ func New(cfg *config.Config, managerClient managerclient.V2) (Announcer, error)
 
 // Started announcer server.
 func (a *announcer) Serve() error {
+	if a.trainerClient != nil {
```
Comment (gaius-qi) [inline scheduler/announcer/announcer.go:108]: Move to announceToManager back.

**Review by gaius-qi — CHANGES_REQUESTED** [review]
